### PR TITLE
[JIT] fix common_expression_hoisting

### LINKED ISSUE #74056: REGR: Accessing dict in JITed code in 1.11
### 🐛 Describe the bug

The following code worked with pytorch 1.10.2. Weirdly, there are many small independent (and sometimes seemingly unrelated) changes that make this error disappear.

```py
from typing import Final

import torch


class LinearActivation(torch.nn.Module):
    def forward(
        self, x: torch.Tensor, meta: dict[str, torch.Tensor]
    ) -> tuple[torch.Tensor, dict[str, torch.Tensor]]:
        meta = meta.copy()
        meta["meta_y_hat"] = x
        # return x, meta # would make it work
        return meta["meta_y_hat"], meta  # JIT claims it errors here


class Test(torch.nn.Module):
    def __init__(self):
        super().__init__()

        self.flag: Final = ""
        self.activation = LinearActivation()

    def forward(
        self, x: torch.Tensor, meta: dict[str, torch.Tensor]
    ) -> tuple[torch.Tensor, dict[str, torch.Tensor]]:
        meta = meta.copy()
        if self.flag != "":  # this branch should not even be compiled
            # assert False # would make it work
            score = x[:, -1:]
            x, meta = self.activation(
                x[:, :, :-1],  # replacing this with x, would make it work
                meta,
            )
            x = torch.cat((x, score), dim=1)  # removing this line makes it work
        else:
            x, meta = self.activation(x, meta)
        meta["meta_y_hat"] = x  # removing this line makes it work
        return meta["meta_y_hat"], meta


if __name__ == "__main__":
    model = torch.jit.script(Test())
    x, xs = model.forward(torch.ones(10, 10), {})
```

```
python test.py
Traceback (most recent call last):
  File "test.py", line 43, in <module>
    x, xs = model.forward(torch.ones(10, 10), {})
RuntimeError: The following operation failed in the TorchScript interpreter.
Traceback of TorchScript (most recent call last):
  File "test.py", line 36, in forward
            x = torch.cat((x, score), dim=1)  # removing this line makes it work
        else:
            x, meta = self.activation(x, meta)
                      ~~~~~~~~~~~~~~~ <--- HERE
        meta["meta_y_hat"] = x  # removing this line makes it work
        return meta["meta_y_hat"], meta
  File "test.py", line 13, in forward
        meta["meta_y_hat"] = x
        # return x, meta # would make it work
        return meta["meta_y_hat"], meta  # JIT claims it errors here
               ~~~~~~~~~~~~~~~~~ <--- HERE
RuntimeError: KeyError: meta_y_hat
```

### Versions

Collecting environment information...
PyTorch version: 1.11.0+cu102
Is debug build: False
CUDA used to build PyTorch: 10.2
ROCM used to build PyTorch: N/A

OS: Ubuntu 16.04.7 LTS (x86_64)
GCC version: (Ubuntu 9.4.0-1ubuntu1~16.04) 9.4.0
Clang version: 3.8.0-2ubuntu4 (tags/RELEASE_380/final)
CMake version: version 3.15.1
Libc version: glibc-2.23

Python version: 3.9.9+ (heads/3.9:86c1265, Nov 27 2021, 15:32:47)  [GCC 9.4.0] (64-bit runtime)
Python platform: Linux-4.13.0-36-generic-x86_64-with-glibc2.23
Is CUDA available: True
CUDA runtime version: 10.2.89
GPU models and configuration: GPU 0: GeForce RTX 2080 Ti
Nvidia driver version: 440.64
cuDNN version: /usr0/local/cuda-9.0/lib64/libcudnn.so.7.0.5
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

Versions of relevant libraries:
[pip3] mypy==0.931
[pip3] mypy-extensions==0.4.3
[pip3] numpy==1.22.3
[pip3] torch==1.11.0+cu102
[pip3] torchvision==0.12.0
[conda] Could not collect

### PR DESCRIPTION
**Pull request by davidberard98**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #74924
* __->__ #74794

aliasdb check was incorrect, caused issues in #74056

This checks whether it is safe to move before the node at the beginning
of the if branch instead of after the node at the beginning of the if
branch

Differential Revision: [D35188927](https://our.internmc.facebook.com/intern/diff/D35188927)

### CODE REVIEW

**Comment by davidberard98** [comment]
@davidberard98 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D35188927).

**Comment by davidberard98** [comment]
Turns out previous fix is insufficient, here's a new repro that still fails

```python
import torch
from typing import Dict

def fn(x, y, cond: bool, d: Dict[str, torch.Tensor]):
    if cond:
        m = x.relu()
        f1 = torch.rand((2, 2))
        d["test"] = f1
        z = d["test"]
    else:
        m = y.gelu()
        f2 = torch.rand((3, 2))
        d["test"] = f2
        z = d["test"]
    return m, z

fn_s = torch.jit.script(fn)
x = torch.rand((2, 2))
y = torch.rand((2, 2))
d = {"x": x}
fn_s(x, y, 1, d)
```

**Comment by Gamrix** [comment]
The fix looks good, but there seems to be a second bug that is unrelated to this change that you are catching with your new repro unless GraphExecutor is running optimizations on the first graph run. 

Print of the graph right before the `fn_s(...)` call in your repro:
```
graph(%x.1 : Tensor,
      %y.1 : Tensor,
      %cond.1 : bool,
      %d.1 : Dict(str, Tensor)):
  %24 : str = prim::Constant[value="none"]()
  %18 : str = prim::Constant[value="test"]() # <ipython-input-2-bf373c45d473>:8:10
  %11 : NoneType = prim::Constant()
  %8 : int = prim::Constant[value=2]() # <ipython-input-2-bf373c45d473>:7:25
  %26 : int = prim::Constant[value=3]() # <ipython-input-2-bf373c45d473>:12:25
  %m : Tensor, %z : Tensor = prim::If(%cond.1) # <ipython-input-2-bf373c45d473>:5:4
    block0():
      %m.1 : Tensor = aten::relu(%x.1) # <ipython-input-2-bf373c45d473>:6:12
      %10 : int[] = prim::ListConstruct(%8, %8)
      %f1.1 : Tensor = aten::rand(%10, %11, %11, %11, %11) # <ipython-input-2-bf373c45d473>:7:13
       = aten::_set_item(%d.1, %18, %f1.1) # <ipython-input-2-bf373c45d473>:8:8
      %z.1 : Tensor = aten::__getitem__(%d.1, %18) # <ipython-input-2-bf373c45d473>:9:12
      -> (%m.1, %z.1)
    block1():
      %m.3 : Tensor = aten::gelu(%y.1, %24) # <ipython-input-2-bf373c45d473>:11:12
      %28 : int[] = prim::ListConstruct(%26, %8)
      %f2.1 : Tensor = aten::rand(%28, %11, %11, %11, %11) # <ipython-input-2-bf373c45d473>:12:13
       = aten::_set_item(%d.1, %18, %f2.1) # <ipython-input-2-bf373c45d473>:13:8
      %z.3 : Tensor = aten::__getitem__(%d.1, %18) # <ipython-input-2-bf373c45d473>:14:12
      -> (%m.3, %z.3)
  %51 : (Tensor, Tensor) = prim::TupleConstruct(%m, %z)
  return (%51)
  ```

**Review by Gamrix — APPROVED** [review]

**Comment by Gamrix** [comment]
From your repo, `aten::_set_item` is somehow not being treated as aliasing  `aten::_getitem__` 

```
[DUMP common_expression_hoisting.cpp:148] After CEH Changes
[DUMP common_expression_hoisting.cpp:148] graph(%x.1 : Tensor,
[DUMP common_expression_hoisting.cpp:148]       %y.1 : Tensor,
[DUMP common_expression_hoisting.cpp:148]       %cond.1 : bool,
[DUMP common_expression_hoisting.cpp:148]       %d.1 : Dict(str, Tensor)):
[DUMP common_expression_hoisting.cpp:148]   %6 : NoneType = prim::Constant()
[DUMP common_expression_hoisting.cpp:148]   %7 : str = prim::Constant[value="test"]() # misc/ceh_bug_repro.py:9:10
[DUMP common_expression_hoisting.cpp:148]   %8 : str = prim::Constant[value="none"]()
[DUMP common_expression_hoisting.cpp:148]   %z.3 : Tensor = aten::__getitem__(%d.1, %7) # misc/ceh_bug_repro.py:15:12
[DUMP common_expression_hoisting.cpp:148]   %m : Tensor = prim::If(%cond.1) # misc/ceh_bug_repro.py:6:4
[DUMP common_expression_hoisting.cpp:148]     block0():
[DUMP common_expression_hoisting.cpp:148]       %m.1 : Tensor = aten::relu(%x.1) # misc/ceh_bug_repro.py:7:12
[DUMP common_expression_hoisting.cpp:148]       %20 : int[] = prim::Constant[value=[2, 2]]()
[DUMP common_expression_hoisting.cpp:148]       %f1.1 : Tensor = aten::rand(%20, %6, %6, %6, %6) # misc/ceh_bug_repro.py:8:13
[DUMP common_expression_hoisting.cpp:148]        = aten::_set_item(%d.1, %7, %f1.1) # misc/ceh_bug_repro.py:9:8
[DUMP common_expression_hoisting.cpp:148]       -> (%m.1)
[DUMP common_expression_hoisting.cpp:148]     block1():
[DUMP common_expression_hoisting.cpp:148]       %m.3 : Tensor = aten::gelu(%y.1, %8) # misc/ceh_bug_repro.py:12:12
[DUMP common_expression_hoisting.cpp:148]       %21 : int[] = prim::Constant[value=[3, 2]]()
[DUMP common_expression_hoisting.cpp:148]       %f2.1 : Tensor = aten::rand(%21, %6, %6, %6, %6) # misc/ceh_bug_repro.py:13:13
[DUMP common_expression_hoisting.cpp:148]        = aten::_set_item(%d.1, %7, %f2.1) # misc/ceh_bug_repro.py:14:8
[DUMP common_expression_hoisting.cpp:148]       -> (%m.3)
[DUMP common_expression_hoisting.cpp:148]   %19 : (Tensor, Tensor) = prim::TupleConstruct(%m, %z.3)
[DUMP common_expression_hoisting.cpp:148]   return (%19)
```

**Comment by davidberard98** [comment]
@Gamrix `couldMoveBeforeTopologicallyValid` returns true if it's possible to move the specified node and some "[working set](https://github.com/pytorch/pytorch/blob/8ed6cb42ba2635319b67417b9ee2f1a0142d2c25/torch/csrc/jit/ir/alias_analysis.cpp#L1708)" to before the target node

In the case of the 2nd repro, we try to move the `getitem` node. It's possible to move this all before the first node as long as we include the `set_item` node in the working set, so `couldMoveBeforeTopologicallyValid` returns true.

So then we were moving `getitem` outside of the if block, even if we require `setattr` to appear before it.

File: torch/csrc/jit/passes/common_expression_hoisting.cpp [hunk 1]
```diff
@@ -42,18 +42,34 @@ struct CommonExpressionHoister {
...
+      TORCH_INTERNAL_ASSERT(
+          did_move_true && did_move_false,
+          "Wasn't able to move nodes to the beginning of the if blocks");
+
+      // Even though we moved the node before the original first node in the
+      // block, we might have also moved other nodes in front of the target
+      // node. If it ended up at the very beginning of the if block, then it's
+      // safe to move it outside.
+      if (true_b_node != true_block->nodes().front() ||
+          false_b_node != false_block->nodes().front()) {
```
[the remark this answers is not in the document]
Comment (davidberard98) [reply torch/csrc/jit/passes/common_expression_hoisting.cpp:69]: I think it's guaranteed that n_{i+1}, ... , n_i are unaffected by this comment https://github.com/pytorch/pytorch/blob/9872a06d77582e91e834103db75f774ca75f7fff/torch/csrc/jit/ir/alias_analysis.h#L115-L123

[the remark this answers is not in the document]
Comment (Gamrix) [reply torch/csrc/jit/passes/common_expression_hoisting.cpp:69]: You're right, I forgot that we already incremented the iterator before we did all the moves.

**Comment by davidberard98** [comment]
@davidberard98 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D35188927).

**Comment by davidberard98** [comment]
@davidberard98 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D35188927).

**Pull request closed without merging**